src/player.c

```diff
@@ -1926,3 +1926,120 @@ void xmp_get_frame_info(xmp_context opaque, struct xmp_frame_info *info)
+		for (pat = 0; pat < mod->pat; pat++) {
+			mod->xxp[pat] = realloc(mod->xxp[pat], sizeof (struct xmp_pattern) + sizeof (int) * (mod->chn + num - 1));
+			if (mod->xxp[pat] == NULL)
+				return -XMP_ERROR_SYSTEM;
+			for (c = 0; c < mod->chn + num; c++) {
+				t = pat * (mod->chn + num) + c;
+				mod->xxp[pat]->index[c] = t;
+				mod->xxt[t] = mod->xxt[pat * mod->chn + c - (c >= start ? num : 0)];
+			}
+		}
+		for (c = start; c < mod->chn + num; c++) {
+			mod->xxc[c].pan = mod->xxc[c - num].pan;
+			mod->xxc[c].vol = mod->xxc[c - num].vol;
+			mod->xxc[c].flg = mod->xxc[c - num].flg;
+		}
+		/* Shrink track matrix */
+		mod->xxt = realloc(mod->xxt, sizeof (struct xmp_track *) * mod->pat * (mod->chn + num));
+		if (mod->xxt == NULL)
+			return -XMP_ERROR_SYSTEM;
+		mod->chn += num;
+		mod->trk = mod->chn * mod->pat;
+	} else {
+		/* See libxmp_init_pattern and, for example, it_load */
+		mod->xxt = realloc(mod->xxt, sizeof (struct xmp_track *) * mod->pat * (mod->chn + num));
+		if (mod->xxt == NULL)
+			return -XMP_ERROR_SYSTEM;
+
+		/* Spread tracks over new channel dimensions */
+		for (pat = mod->pat - 1; pat >= 0; pat--) {
+			for (c = mod->chn - 1; c >= 0; c--) {
+				mod->xxt[pat * (mod->chn + num) + c] = mod->xxt[pat * mod->chn + c];
+			}
+		}
+
+		mod->chn += num;
+		mod->trk = mod->chn * mod->pat;
+
+		/* Shift tracks forward */
+		for (pat = mod->pat - 1; pat >= 0; pat--) {
+			mod->xxp[pat] = realloc(mod->xxp[pat], sizeof (struct xmp_pattern) + sizeof (int) * mod->chn - 1);
+			if (mod->xxp[pat] == NULL)
+				return -XMP_ERROR_SYSTEM;
+			for (c = mod->chn - 1; c >= 0; c--) {
+				t = pat * mod->chn + c;
+				mod->xxp[pat]->index[c] = t;
+				if (c >= start + num) {
+					mod->xxt[t] = mod->xxt[t - num];
+				} else if (c >= start) {
+					/* See libxmp_alloc_track */
+					mod->xxt[t] = calloc(sizeof (struct xmp_track) + sizeof (struct xmp_event) * (mod->xxp[pat]->rows - 1), 1);
+					if (mod->xxt[t] == NULL)
+						return -XMP_ERROR_SYSTEM;
+					mod->xxt[t]->rows = mod->xxp[pat]->rows;
+				}
+			}
+		}
+
+		for (c = mod->chn - 1; c >= start + num; c--) {
+			mod->xxc[c].pan = mod->xxc[c - num].pan;
+			mod->xxc[c].vol = mod->xxc[c - num].vol;
+			mod->xxc[c].flg = mod->xxc[c - num].flg;
+			/* See xmp_start_player */
+			p->channel_mute[c] = 0;
+			p->channel_vol[c] = 100;
+		}
+		for (c = start + num - 1; c >= start; c--) {
+			/* See libxmp_load_prologue */
+			mod->xxc[c].pan = 0x80;
+			mod->xxc[c].vol = 0x40;
+			mod->xxc[c].flg = 0;
+		}
+
+		/* This is needed to get actual speaker output when in instrument mode! */
+		libxmp_virt_off(ctx);
+		if (libxmp_virt_on(ctx, mod->trk) != 0)
+			return -XMP_ERROR_INTERNAL;
+		/* See xmp_start_player */
+		//p->xc_data = realloc(p->xc_data, p->virt.virt_channels * sizeof (struct channel_data));
+		//if (p->xc_data == NULL)
+		//	return -XMP_ERROR_SYSTEM;
+	}
+
+	libxmp_virt_off(ctx);
+	/* See xmp_play_frame */
+	//libxmp_virt_reset(ctx);
+	//reset_channels(ctx);
```

Comment:
I *think* an `xmp_end_player` call here might be sufficient to reset virtual channels/`p->xc_data`/`f->loop` but it would require that the user manually calls `xmp_start_player` again (since the rate and format arguments do not seem to be stored). Alternatively, if this is being allowed in the playing state without stopping the playing state, I think `libxmp_virt_reset` and `reset_channels` might be correct (honestly, I'm not sure and I'm still figuring out how this part of libxmp works).

Reply:
Your suspicions are correct.  Testing was performed in real time while a module played.  If I remember correctly, `reset_channels` was required to prevent playback from being interrupted.

More importantly, however, I apologize for not making it clearer [here](https://github.com/libxmp/libxmp/pull/274#issuecomment-753867642) that 9e74e75 is the commit to evaluate, as subsequent commits basically comprise experimentation involving Travis, as I am currently unable to reproduce the original leak locally.  This includes the commented-out lines here (and elsewhere, too, of course).